### PR TITLE
Severity monoid

### PR DESCRIPTION
**Pull request by coot**
This patch introduces a commutative monoid instance for Severity which
should be used for any severity calucations.  Unfortuntately it is not
possible to remove `Ord` instance, which can be confused.  This patch
also fixes `testSeverity` computation.  It should compare the message
severity with minimum severity found in any configuration settings.
This allows to turn on messages from a tracer  for better dubugging
information, rather than turn them off.  That seems like a more
sensible default.

### CODE REVIEW

**Comment by coot** [comment]
Are any other places where using the monoid instance would fit its purpose?

Removing `Ord` instance would be nice, but this might be difficult or even impossible.

**Comment by CodiePP** [comment]
> Are any other places where using the monoid instance would fit its purpose?
> 
> Removing `Ord` instance would be nice, but this might be difficult or even impossible.

we need to be able to compare two severities. how else could it be done?

File: iohk-monitoring/src/Cardano/BM/Data/Severity.lhs
Comment (mrBliss) [inline iohk-monitoring/src/Cardano/BM/Data/Severity.lhs:63]: So `Debug <> Emergency = Debug`?

This is not intuitive to me at all. Say I get a bunch of messages with each a severity, I want them to be traced with the *highest* severity of the bunch, right? Analogy: when you send 3 letters in the same envelope, the priority of the envelope is determined by the letter with the highest priority.

`(<>)` = `max` would make much more sense to me.

**Comment by CodiePP** [comment]
changing the intented behaviour of this test might confuse people. could there be another possiblity? like using "negative" severity limit to indicate that this should overwrite the global minSeverity?

**Comment by coot** [comment]
For any lower semi lattice (which a `Severity` is) you can compare using: `a <> b == a`, which is `True` iff `a <= b`.

**Comment by CodiePP** [comment]
apart from that I think the monoid approach is nice!

**Comment by coot** [comment]
> apart from that I think the monoid approach is nice!

Yes it makes it more difficult to mistake `min` with `max`.  Another option is to provide an `Ord` instance through a newtype wrapper:

```
newtype SeverityOrd = SeverityOrd { getSeverity :: Severity }

instance Ord (SeverityOrd) where
  (SeverityOrd a) `compare` (SeveriyttOrd b) =
    if | a == b -> EQ
       | a <> b == a -> LT
       | otherwise = GT
```

File: iohk-monitoring/src/Cardano/BM/Data/Severity.lhs
Comment (coot) [reply iohk-monitoring/src/Cardano/BM/Data/Severity.lhs]: At the top I argue for `min`:

>  It should compare the message severity with minimum severity found in any configuration settings. This allows to turn on messages from a tracer for better dubugging information, rather than turn them off. That seems like a more sensible default.

File: iohk-monitoring/src/Cardano/BM/Data/Severity.lhs
Comment (mrBliss) [inline iohk-monitoring/src/Cardano/BM/Data/Severity.lhs:77]: So this is just a convoluted way to say `(<>) = min`? :grin:

File: iohk-monitoring/src/Cardano/BM/Configuration.lhs
Comment (mrBliss) [inline iohk-monitoring/src/Cardano/BM/Configuration.lhs:70]: Hmm, you argued that `<>` should be `min`, but here the previous code was clearly `max`, right?

Comment (coot) [reply iohk-monitoring/src/Cardano/BM/Configuration.lhs:70]: Indeed, this was a :bug:

File: iohk-monitoring/src/Cardano/BM/Data/Severity.lhs
Comment (coot) [reply iohk-monitoring/src/Cardano/BM/Data/Severity.lhs:77]: Yes, I just want to encourage to remove the `Ord` instance.

**Comment by mrBliss** [comment]
> Yes, I just want to encourage to remove the `Ord` instance.

I agree, because I just realised I had flipped the order in my head :facepalm:.

File: iohk-monitoring/src/Cardano/BM/Data/Severity.lhs
Comment (mrBliss) [reply iohk-monitoring/src/Cardano/BM/Data/Severity.lhs]: I see, okay, for combining severities from configs, `min` makes sense (and is confusing, so `(<>)` is indeed better).

File: iohk-monitoring/src/Cardano/BM/Data/Severity.lhs
Comment (dcoutts) [inline iohk-monitoring/src/Cardano/BM/Data/Severity.lhs:77]: I agree with adding the monoid instance.

I'm not sure I agree that we could remove ord. We could only remove Ord if we provide and use threshold-testing functions everywhere.

**Review by dcoutts — APPROVED** [review]

File: iohk-monitoring/src/Cardano/BM/Data/Severity.lhs
Comment (coot) [reply iohk-monitoring/src/Cardano/BM/Data/Severity.lhs:77]: > I'm not sure I agree that we could remove ord. We could only remove Ord if we provide and use threshold-testing functions everywhere.

If it is too difficult to remove it, because it appears in other types that need to have `Ord` instances or otherwise,  we could use something like [Ordered](https://hackage.haskell.org/package/lattices-2.0.2/docs/Algebra-Lattice-Ordered.html#t:Ordered) newtype wrapper.  At least one needs to be explicit to use the `Ord` instance.

Comment (deepfire) [reply iohk-monitoring/src/Cardano/BM/Data/Severity.lhs:77]: I'm somewhat confused about what makes `Ord` confusing -- and what would make the `Ordered` less confusing?

`Ord` provides an ordering, that is used by `>`/`<`.  

`Ordering` would also provide an ordering, and would also come together with functions with similar meaning.

Where is the explanatory benefit?

Comment (coot) [reply iohk-monitoring/src/Cardano/BM/Data/Severity.lhs:77]: Threshold-testing can be done in this way:
```
atLeastWarning a = a <> Warning == Warning
```
no need for `Ord` to do that.  By the way, in lattice theory this the way to show that every semi lattice is a partial order (with finite infima) and vice versa if I recall correctly.

**Comment by coot** [comment]
I think I need one more review from one the maintainers.

**Review by deepfire — APPROVED** [review]
bors r+

Thank you!

**Comment by deepfire** [comment]
This now gives a test failure: https://iohk-nix-cache.s3-eu-central-1.amazonaws.com/log/3wc3wmzbwf6zv61zqv26x21zba59balq-iohk-monitoring-test-tests-0.1.10.1-check.drv

**Comment by deepfire** [comment]
Ok, https://github.com/input-output-hk/iohk-monitoring-framework/pull/598 now exposes the CI failure.

**Comment by deepfire** [comment]
So it looks like we've replaced what used to be `fromMaybe Debug globnamesev` with `fold globnamesev` which is equivalent to `fromMaybe Emergency globnamesev`, because of the `Monoid` having `mempty` at `Evergency`.

So that's not as non-confusing, as it appears to be, given that it passed review : -)

**Comment by deepfire** [comment]
So it looks like we've replaced what used to be `fromMaybe Debug globnamesev` with `fold globnamesev` which is equivalent to `fromMaybe Emergency globnamesev`, because of the `Monoid` having `mempty` at `Emergency`.

So that's not as non-confusing, as it appears to be, given that it passed review : -)

**Comment by mrBliss** [comment]
> So it looks like we've replaced what used to be `fromMaybe Debug globnamesev` with `fold globnamesev` which is equivalent to `fromMaybe Emergency globnamesev`, because of the `Monoid` having `mempty` at `Emergency`.
> 
> So that's not as non-confusing, as it appears to be, given that it passed review : -)

(FYI, https://github.com/input-output-hk/iohk-monitoring-framework/pull/593/files#r510283127)

**Pull request merged**
